reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,160 @@
+uses resources restricted to each spark job allocated resources.
+
+
+### Key questions
+
+#### We use spark barrier mode job to launch a ray cluster, shall we make the spark barrier to be a background job and keep running, until user explicitly terminate it ?
+
+Yes. So every user only need to start ray cluster once and then user can run ray applications
+on the ray cluster with little overhead.
+
+
+#### Launch Ray head node on spark driver node or spark task side ?
+Because we need to allocate compute resources to the ray head node, If launching in spark task side,
+it might consume driver side resources, and if there are too many users sharing the spark cluster,
+spark driver resources might be exhausted.
+So, I suggest to launch Ray head node on spark task side, we can pick the first spark task to launch
+head node.
+Issue: How to send the ray head node IP / port to spark driver side ?
+So that on spark driver side user can run  the ray application on the ray cluster.
+Because the spark barrier job keeps running when the ray cluster is up, we cannot use
+`rdd.collect()` to send the head node IP / port  back to driver side, we need to broadcast
+the spark driver IP to task side and then in spark task, we can send the  head node IP / port
+back to driver node by TCP connection.
+
+
+#### How to select the port number used by Ray node ?
+Ray node requires listening on several port, a spark cluster might be shared by many users,
+each users might setup their own ray cluster concurrently, to avoid port conflicts,
+we can randomly select free port and start Ray node service,
+if failed we can retry on another free port.
+
+
+#### How much memory shall we allocate to Ray node service ( set via ray script --memory option) ?
+Spark does not provide explicit API for getting task allowed memory,
+I propose:
+SPARK_WORKER_MEMORY / SPARK_WORKER_CORES * RAY_NODE_NUM_CPUS
+But how to get SPARK_WORKER_MEMORY and SPARK_WORKER_CORES in spark task is an issue.
+
+
+#### How to make ray respect spark GPU resource scheduling ?
+In spark task, we can get GPU IDs allocated to this task, so, when launching
+Ray node, besides specifying `--num-gpus` options, we need specify `CUDA_VISIBLE_DEVICES`
+environment so that we can restrict Ray node only uses the GPUs allocated to corresponding spark tasks.
+
+
+### API Proposal
+
+#### Initialize ray cluster on spark API
+
+```
+ray_cluster_address, ray_cluster_handler = ray.spark.init_cluster(num_cpus, num_gpus, memory)
+```
+
+Init a ray cluster on the spark cluster, the argumens specified the ray cluster can use how much cpus / gpus / memory.
+
+
+Returns a `ray_cluster_address` string (Ray Head node IP / port) and a `ray_cluster_handler`
+
+
+#### Initialize ray client API.
+
+Now User can run ray application on the returned ray cluster, by executing `ray.init(address=ray_cluster_address)`, and then run any ray application code.
```

Comment:
It would be preferable to avoid Ray client (use Ray direct connection), at least by default. This is for improved performance with data transferred to the Ray driver.

Reply:
> It would be preferable to avoid Ray client (use Ray direct connection)

Do you have example code or document link about this ? :)

Reply:
If `ray start` has been run on a node, then `ray.init()` will automatically discover and connect to the raylet on that node. This works on any node (head or worker).

Reply:
> If ray start has been run on a node, then ray.init() will automatically discover and connect to the raylet on that node. 

Sounds good!
Would you tell me how it is implemented ?

Reply:
The Ray cluster is discovered by checking the file `/tmp/ray/ray_current_cluster`

Reply:
@ericl 
Issue: If the Ray cluster is discovered by checking the file /tmp/ray/ray_current_cluster,
suppose in a shared spark cluster, multiple Ray cluster is launched (by different users), and Ray cluster head nodes are all launched in spark driver node, then they share the same `/tmp/ray/ray_current_cluster` file, how to make each user connect to their own ray cluster ?

Reply:
If there are multiple Ray clusters, then you'd have to implement discovery manually. This could be done by tracking the gcs port of the head node, and then passing that port explicitly to `ray.init(address=ip:port)`.

Reply:
> tracking the gcs port of the head node, and then passing that port explicitly to ray.init(address=ip:port).

yep, this is the approach used in my prototype code